// @mui/x-date-pickers/legacy/DateTimePicker/DateTimePickerTabs.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import * as React from 'react';
import PropTypes from 'prop-types';
import Tab from '@mui/material/Tab';
import Tabs, { tabsClasses } from '@mui/material/Tabs';
import { styled, useThemeProps } from '@mui/material/styles';
import { unstable_composeClasses as composeClasses } from '@mui/utils';
import { Time, DateRange } from '../internals/components/icons';
import { useLocaleText } from '../internals/hooks/useUtils';
import { getDateTimePickerTabsUtilityClass } from './dateTimePickerTabsClasses';
import { jsx as _jsx } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";
var viewToTab = function viewToTab(view) {
  if (['day', 'month', 'year'].includes(view)) {
    return 'date';
  }
  return 'time';
};
var tabToView = function tabToView(tab) {
  if (tab === 'date') {
    return 'day';
  }
  return 'hours';
};
var useUtilityClasses = function useUtilityClasses(ownerState) {
  var classes = ownerState.classes;
  var slots = {
    root: ['root']
  };
  return composeClasses(slots, getDateTimePickerTabsUtilityClass, classes);
};
var DateTimePickerTabsRoot = styled(Tabs, {
  name: 'MuiDateTimePickerTabs',
  slot: 'Root',
  overridesResolver: function overridesResolver(_, styles) {
    return styles.root;
  }
})(function (_ref) {
  var theme = _ref.theme;
  return {
    boxShadow: "0 -1px 0 0 inset ".concat((theme.vars || theme).palette.divider),
    '&:last-child': _defineProperty({
      boxShadow: "0 1px 0 0 inset ".concat((theme.vars || theme).palette.divider)
    }, "& .".concat(tabsClasses.indicator), {
      bottom: 'auto',
      top: 0
    })
  };
});
var DateTimePickerTabs = function DateTimePickerTabs(inProps) {
  var props = useThemeProps({
    props: inProps,
    name: 'MuiDateTimePickerTabs'
  });
  var _props$dateIcon = props.dateIcon,
    dateIcon = _props$dateIcon === void 0 ? /*#__PURE__*/_jsx(DateRange, {}) : _props$dateIcon,
    onViewChange = props.onViewChange,
    _props$timeIcon = props.timeIcon,
    timeIcon = _props$timeIcon === void 0 ? /*#__PURE__*/_jsx(Time, {}) : _props$timeIcon,
    view = props.view,
    _props$hidden = props.hidden,
    hidden = _props$hidden === void 0 ? typeof window === 'undefined' || window.innerHeight < 667 : _props$hidden;
  var localeText = useLocaleText();
  var classes = useUtilityClasses(props);
  var handleChange = function handleChange(event, value) {
    onViewChange(tabToView(value));
  };
  if (hidden) {
    return null;
  }
  return /*#__PURE__*/_jsxs(DateTimePickerTabsRoot, {
    ownerState: props,
    variant: "fullWidth",
    value: viewToTab(view),
    onChange: handleChange,
    className: classes.root,
    children: [/*#__PURE__*/_jsx(Tab, {
      value: "date",
      "aria-label": localeText.dateTableLabel,
      icon: /*#__PURE__*/_jsx(React.Fragment, {
        children: dateIcon
      })
    }), /*#__PURE__*/_jsx(Tab, {
      value: "time",
      "aria-label": localeText.timeTableLabel,
      icon: /*#__PURE__*/_jsx(React.Fragment, {
        children: timeIcon
      })
    })]
  });
};
process.env.NODE_ENV !== "production" ? DateTimePickerTabs.propTypes = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // | To update them edit the TypeScript types and run "yarn proptypes"  |
  // ----------------------------------------------------------------------
  /**
   * Override or extend the styles applied to the component.
   */
  classes: PropTypes.object,
  /**
   * Date tab icon.
   * @default DateRange
   */
  dateIcon: PropTypes.node,
  /**
   * Toggles visibility of the tabs allowing view switching.
   * @default `window.innerHeight < 667` for `DesktopDateTimePicker` and `MobileDateTimePicker`, `displayStaticWrapperAs === 'desktop'` for `StaticDateTimePicker`
   */
  hidden: PropTypes.bool,
  /**
   * Callback called when a tab is clicked
   * @template TView
   * @param {TView} view The view to open
   */
  onViewChange: PropTypes.func.isRequired,
  /**
   * Time tab icon.
   * @default Time
   */
  timeIcon: PropTypes.node,
  /**
   * Currently visible picker view.
   */
  view: PropTypes.oneOf(['day', 'hours', 'minutes', 'month', 'seconds', 'year']).isRequired
} : void 0;
export { DateTimePickerTabs };
