// @mui/x-date-pickers/node/internals/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "DAY_MARGIN", {
  enumerable: true,
  get: function () {
    return _dimensions.DAY_MARGIN;
  }
});
Object.defineProperty(exports, "DEFAULT_DESKTOP_MODE_MEDIA_QUERY", {
  enumerable: true,
  get: function () {
    return _utils.DEFAULT_DESKTOP_MODE_MEDIA_QUERY;
  }
});
Object.defineProperty(exports, "DIALOG_WIDTH", {
  enumerable: true,
  get: function () {
    return _dimensions.DIALOG_WIDTH;
  }
});
Object.defineProperty(exports, "DayCalendar", {
  enumerable: true,
  get: function () {
    return _DayCalendar.DayCalendar;
  }
});
Object.defineProperty(exports, "PickersArrowSwitcher", {
  enumerable: true,
  get: function () {
    return _PickersArrowSwitcher.PickersArrowSwitcher;
  }
});
Object.defineProperty(exports, "PickersCalendarHeader", {
  enumerable: true,
  get: function () {
    return _PickersCalendarHeader.PickersCalendarHeader;
  }
});
Object.defineProperty(exports, "PickersModalDialog", {
  enumerable: true,
  get: function () {
    return _PickersModalDialog.PickersModalDialog;
  }
});
Object.defineProperty(exports, "PickersPopper", {
  enumerable: true,
  get: function () {
    return _PickersPopper.PickersPopper;
  }
});
Object.defineProperty(exports, "PickersToolbar", {
  enumerable: true,
  get: function () {
    return _PickersToolbar.PickersToolbar;
  }
});
Object.defineProperty(exports, "PickersToolbarButton", {
  enumerable: true,
  get: function () {
    return _PickersToolbarButton.PickersToolbarButton;
  }
});
Object.defineProperty(exports, "addPositionPropertiesToSections", {
  enumerable: true,
  get: function () {
    return _useField.addPositionPropertiesToSections;
  }
});
Object.defineProperty(exports, "applyDefaultDate", {
  enumerable: true,
  get: function () {
    return _dateUtils.applyDefaultDate;
  }
});
Object.defineProperty(exports, "areDatesEqual", {
  enumerable: true,
  get: function () {
    return _dateUtils.areDatesEqual;
  }
});
Object.defineProperty(exports, "buildDeprecatedPropsWarning", {
  enumerable: true,
  get: function () {
    return _warning.buildDeprecatedPropsWarning;
  }
});
Object.defineProperty(exports, "buildWarning", {
  enumerable: true,
  get: function () {
    return _warning.buildWarning;
  }
});
Object.defineProperty(exports, "createDateStrForInputFromSections", {
  enumerable: true,
  get: function () {
    return _useField.createDateStrForInputFromSections;
  }
});
Object.defineProperty(exports, "defaultReduceAnimations", {
  enumerable: true,
  get: function () {
    return _defaultReduceAnimations.defaultReduceAnimations;
  }
});
Object.defineProperty(exports, "executeInTheNextEventLoopTick", {
  enumerable: true,
  get: function () {
    return _utils.executeInTheNextEventLoopTick;
  }
});
Object.defineProperty(exports, "extractValidationProps", {
  enumerable: true,
  get: function () {
    return _extractValidationProps.extractValidationProps;
  }
});
Object.defineProperty(exports, "getActiveElement", {
  enumerable: true,
  get: function () {
    return _utils.getActiveElement;
  }
});
Object.defineProperty(exports, "onSpaceOrEnter", {
  enumerable: true,
  get: function () {
    return _utils.onSpaceOrEnter;
  }
});
Object.defineProperty(exports, "pickersArrowSwitcherClasses", {
  enumerable: true,
  get: function () {
    return _pickersArrowSwitcherClasses.pickersArrowSwitcherClasses;
  }
});
Object.defineProperty(exports, "pickersPopperClasses", {
  enumerable: true,
  get: function () {
    return _pickersPopperClasses.pickersPopperClasses;
  }
});
Object.defineProperty(exports, "pickersToolbarButtonClasses", {
  enumerable: true,
  get: function () {
    return _pickersToolbarButtonClasses.pickersToolbarButtonClasses;
  }
});
Object.defineProperty(exports, "pickersToolbarClasses", {
  enumerable: true,
  get: function () {
    return _pickersToolbarClasses.pickersToolbarClasses;
  }
});
Object.defineProperty(exports, "pickersToolbarTextClasses", {
  enumerable: true,
  get: function () {
    return _pickersToolbarTextClasses.pickersToolbarTextClasses;
  }
});
Object.defineProperty(exports, "replaceInvalidDateByNull", {
  enumerable: true,
  get: function () {
    return _dateUtils.replaceInvalidDateByNull;
  }
});
Object.defineProperty(exports, "splitFieldInternalAndForwardedProps", {
  enumerable: true,
  get: function () {
    return _fields.splitFieldInternalAndForwardedProps;
  }
});
Object.defineProperty(exports, "uncapitalizeObjectKeys", {
  enumerable: true,
  get: function () {
    return _slotsMigration.uncapitalizeObjectKeys;
  }
});
Object.defineProperty(exports, "useCalendarState", {
  enumerable: true,
  get: function () {
    return _useCalendarState.useCalendarState;
  }
});
Object.defineProperty(exports, "useDefaultDates", {
  enumerable: true,
  get: function () {
    return _useUtils.useDefaultDates;
  }
});
Object.defineProperty(exports, "useField", {
  enumerable: true,
  get: function () {
    return _useField.useField;
  }
});
Object.defineProperty(exports, "useLocaleText", {
  enumerable: true,
  get: function () {
    return _useUtils.useLocaleText;
  }
});
Object.defineProperty(exports, "useLocalizationContext", {
  enumerable: true,
  get: function () {
    return _useUtils.useLocalizationContext;
  }
});
Object.defineProperty(exports, "useNextMonthDisabled", {
  enumerable: true,
  get: function () {
    return _dateHelpersHooks.useNextMonthDisabled;
  }
});
Object.defineProperty(exports, "useNow", {
  enumerable: true,
  get: function () {
    return _useUtils.useNow;
  }
});
Object.defineProperty(exports, "usePicker", {
  enumerable: true,
  get: function () {
    return _usePicker.usePicker;
  }
});
Object.defineProperty(exports, "usePreviousMonthDisabled", {
  enumerable: true,
  get: function () {
    return _dateHelpersHooks.usePreviousMonthDisabled;
  }
});
Object.defineProperty(exports, "useStaticPicker", {
  enumerable: true,
  get: function () {
    return _useStaticPicker.useStaticPicker;
  }
});
Object.defineProperty(exports, "useUtils", {
  enumerable: true,
  get: function () {
    return _useUtils.useUtils;
  }
});
Object.defineProperty(exports, "useValidation", {
  enumerable: true,
  get: function () {
    return _useValidation.useValidation;
  }
});
Object.defineProperty(exports, "validateDate", {
  enumerable: true,
  get: function () {
    return _validateDate.validateDate;
  }
});
Object.defineProperty(exports, "validateDateTime", {
  enumerable: true,
  get: function () {
    return _validateDateTime.validateDateTime;
  }
});
Object.defineProperty(exports, "validateTime", {
  enumerable: true,
  get: function () {
    return _validateTime.validateTime;
  }
});
var _PickersArrowSwitcher = require("./components/PickersArrowSwitcher/PickersArrowSwitcher");
var _PickersModalDialog = require("./components/PickersModalDialog");
var _PickersPopper = require("./components/PickersPopper");
var _PickersToolbar = require("./components/PickersToolbar");
var _pickersToolbarClasses = require("./components/pickersToolbarClasses");
var _pickersToolbarButtonClasses = require("./components/pickersToolbarButtonClasses");
var _pickersToolbarTextClasses = require("./components/pickersToolbarTextClasses");
var _pickersArrowSwitcherClasses = require("./components/PickersArrowSwitcher/pickersArrowSwitcherClasses");
var _pickersPopperClasses = require("./components/pickersPopperClasses");
var _PickersToolbarButton = require("./components/PickersToolbarButton");
var _dimensions = require("./constants/dimensions");
var _useField = require("./hooks/useField");
var _usePicker = require("./hooks/usePicker");
var _useStaticPicker = require("./hooks/useStaticPicker");
var _useUtils = require("./hooks/useUtils");
var _useValidation = require("./hooks/useValidation");
var _dateHelpersHooks = require("./hooks/date-helpers-hooks");
var _dateUtils = require("./utils/date-utils");
var _fields = require("./utils/fields");
var _utils = require("./utils/utils");
var _defaultReduceAnimations = require("./utils/defaultReduceAnimations");
var _extractValidationProps = require("./utils/validation/extractValidationProps");
var _validateDate = require("./utils/validation/validateDate");
var _validateDateTime = require("./utils/validation/validateDateTime");
var _validateTime = require("./utils/validation/validateTime");
var _warning = require("./utils/warning");
var _slotsMigration = require("./utils/slots-migration");
var _PickersCalendarHeader = require("../DateCalendar/PickersCalendarHeader");
var _DayCalendar = require("../DateCalendar/DayCalendar");
var _useCalendarState = require("../DateCalendar/useCalendarState");
