// @mui/x-date-pickers/modern/tests/describeValue/describeValue.js
import _extends from "@babel/runtime/helpers/esm/extends";
import * as React from 'react';
import createDescribe from '@mui/monorepo/test/utils/createDescribe';
import { buildFieldInteractions } from 'test/utils/pickers-utils';
import { testControlledUnControlled } from './testControlledUnControlled';
import { testPickerOpenCloseLifeCycle } from './testPickerOpenCloseLifeCycle';
import { testPickerActionBar } from './testPickerActionBar';
import { jsx as _jsx } from "react/jsx-runtime";
const TEST_SUITES = [testControlledUnControlled, testPickerOpenCloseLifeCycle, testPickerActionBar];
function innerDescribeValue(ElementToTest, getOptions) {
  const options = getOptions();
  const {
    defaultProps,
    render,
    clock,
    componentFamily
  } = options;
  function WrappedElementToTest(props) {
    return /*#__PURE__*/_jsx(ElementToTest, _extends({}, defaultProps, props));
  }
  let renderWithProps;
  if (componentFamily === 'field' || componentFamily === 'picker') {
    const interactions = buildFieldInteractions({
      clock,
      render,
      Component: ElementToTest
    });
    renderWithProps = props => interactions.renderWithProps(_extends({}, defaultProps, props), componentFamily);
  } else {
    renderWithProps = props => {
      const response = render( /*#__PURE__*/_jsx(WrappedElementToTest, _extends({}, props)));
      return _extends({}, response, {
        input: null,
        selectSection: () => {
          throw new Error('You can only select a section on components that render a field');
        }
      });
    };
  }
  TEST_SUITES.forEach(testSuite => {
    testSuite(WrappedElementToTest, _extends({}, options, {
      renderWithProps
    }));
  });
}
/**
 * Tests various aspects of component value.
 */
export const describeValue = createDescribe('Value API', innerDescribeValue);
